src/geff/affine.py

```diff
@@ -0,0 +1,168 @@
+    The transformation matrix follows scipy.ndimage.affine_transform convention
+    where the matrix maps output coordinates to input coordinates (inverse/pull transformation).
+
+    For a point p_out in output space, the corresponding input point p_in is computed as:
+    p_in_homo = matrix @ p_out_homo
+    where p_out_homo = [p_out; 1] and p_in = p_in_homo[:-1]
+
+    Attributes:
+        matrix: Homogeneous transformation matrix as list of lists (ndim+1, ndim+1)
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    matrix: Any = Field(
+        ..., description="Homogeneous transformation matrix as list of lists (ndim+1, ndim+1)"
+    )
+
+    @model_validator(mode="after")
+    def validate_matrix(self) -> Affine:
+        """Validate that matrix is a proper homogeneous transformation matrix.
+
+        Converts to list of lists format.
+        """
+        # Convert input to numpy array for validation
+        try:
+            matrix_array = np.asarray(self.matrix, dtype=float)
+        except (ValueError, TypeError) as e:
+            raise ValueError(f"Matrix must be convertible to numpy array: {e}") from e
+
+        if matrix_array.ndim != 2:
+            raise ValueError(f"Matrix must be 2D, got {matrix_array.ndim}D")
+
+        if matrix_array.shape[0] != matrix_array.shape[1]:
+            raise ValueError(f"Matrix must be square, got shape {matrix_array.shape}")
+
+        if matrix_array.shape[0] < 2:
+            raise ValueError(
+                f"Matrix must be at least 2x2 for homogeneous coordinates, got {matrix_array.shape}"
+            )
+
+        # Check that bottom row is [0, 0, ..., 1]
+        expected_bottom_row = np.zeros(matrix_array.shape[1])
+        expected_bottom_row[-1] = 1.0
+
+        if not np.allclose(matrix_array[-1, :], expected_bottom_row):
+            raise ValueError(
+                f"Bottom row of homogeneous matrix must be [0, 0, ..., 1], "
+                f"got {matrix_array[-1, :]}"
+            )
+
+        # Convert to list of lists for storage
+        self.matrix = matrix_array.tolist()
+        return self
+
+    def numpy(self) -> NDArray[np.floating]:
+        """Convert the stored list of lists matrix to numpy array."""
+        return np.array(self.matrix, dtype=float)
+
+    @property
+    def ndim(self) -> int:
+        """Number of spatial dimensions (excluding homogeneous coordinate)."""
+        return len(self.matrix) - 1
+
+    @property
+    def linear_matrix(self) -> NDArray[np.floating]:
+        """Extract the linear transformation part (ndim, ndim)."""
+        matrix_array = self.numpy()
+        return matrix_array[:-1, :-1].copy()
+
+    @property
+    def offset(self) -> NDArray[np.floating]:
+        """Extract the translation offset (ndim,)."""
+        matrix_array = self.numpy()
+        return matrix_array[:-1, -1].copy()
+
+    def transform_points(self, points: NDArray[np.floating]) -> NDArray[np.floating]:
```

Comment:
I'm still on the fence about including this as a feature. Let's get some other opinions.